yazi-plugin/src/external/highlighter.rs

```diff
@@ -56,51 +59,235 @@ impl Highlighter {
 		syntaxes.find_syntax_by_first_line(&line).ok_or_else(|| anyhow!("No syntax found"))
 	}
 
-	pub async fn highlight(&self, skip: usize, limit: usize) -> Result<Text<'static>, PeekError> {
-		let mut reader = BufReader::new(File::open(&self.path).await?);
+	pub async fn highlight(
+		&self,
+		skip: usize,
+		area: Rect,
+		wrap: bool,
+	) -> Result<Text<'static>, PeekError> {
+		let reader = BufReader::new(File::open(&self.path).await?);
 
 		let syntax = Self::find_syntax(&self.path).await;
 		let mut plain = syntax.is_err();
 
-		let mut before = Vec::with_capacity(if plain { 0 } else { skip });
-		let mut after = Vec::with_capacity(limit);
+		let (before, after): (Vec<String>, Vec<String>);
+		if wrap {
+			(before, after) = Self::before_after_wrapped(skip, area, &mut plain, reader).await?;
+		} else {
+			(before, after) = Self::before_after(skip, area, &mut plain, reader).await?;
+		}
 
-		let mut i = 0;
+		if plain {
+			if !wrap {
+				let indent = " ".repeat(PREVIEW.tab_size as usize);
+				return Ok(Text::from(after.join("").replace('\t', &indent)));
+			}
+			Ok(Text::from(after.join("")))
+		} else {
+			Self::highlight_with(before, after, syntax.unwrap()).await
+		}
+	}
+
+	async fn before_after_wrapped(
+		skip: usize,
+		area: Rect,
+		plain: &mut bool,
+		mut reader: BufReader<File>,
+	) -> Result<(Vec<String>, Vec<String>), PeekError> {
+		let mut before = Vec::with_capacity(if *plain { 0 } else { skip });
+		let mut after = Vec::with_capacity(area.height as usize);
+
+		let mut long_lines = vec![];
 		let mut buf = vec![];
+		// If we want to indent plain text, we have to decide if it is plain
 		while reader.read_until(b'\n', &mut buf).await.is_ok() {
-			i += 1;
-			if buf.is_empty() || i > skip + limit {
+			if buf.is_empty() {
 				break;
 			}
-
-			if !plain && buf.len() > 6000 {
-				plain = true;
-				drop(mem::take(&mut before));
+			if !*plain && buf.len() > MAX_LINE_BYTES_TO_PLAINTEXT_FALLBACK {
+				*plain = true;
 			}
+			long_lines.push(buf.clone());
+			buf.clear()
+		}
 
-			if buf.ends_with(b"\r\n") {
-				buf.pop();
-				buf.pop();
-				buf.push(b'\n');
+		let mut lines_handled = 0;
+		for mut long_line in long_lines {
+			if long_line.is_empty() || lines_handled > skip + area.height as usize {
+				break;
+			}
+			Self::replace_tabs_with_spaces(&mut long_line, PREVIEW.tab_size as usize);
+			for line in Self::chunk_by_width(long_line, area.width as usize) {
+				let mut line = line.to_vec();
```

Comment:
```suggestion
			for mut line in Self::chunk_by_width(long_line, area.width as usize) {
```

This can avoid unnecessary memory reallocations.

Reply:
Done